### ISSUE TITLE
Analyzer: Some correct super-bounded types are rejected.

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#34560]
Consider the following program:

```dart
class A<X extends num> {}

A<dynamic> a;
List<A<dynamic>> lst;

main() {
  A<dynamic>(); //# 01: compile-time error
}
```

Storing the above in 'n032.dart' and running dartanalyzer version 2.1.0-dev.5.0, I get the following response:

```
Analyzing n032.dart...
  error • 'dynamic' doesn't extend 'num' at n032.dart:3:3 • type_argument_not_matching_bounds
  error • 'dynamic' doesn't extend 'num' at n032.dart:4:8 • type_argument_not_matching_bounds
  error • 'dynamic' doesn't extend 'num' at n032.dart:7:5 • type_argument_not_matching_bounds
```

However, lines 3 and 4 are correct (it is allowed to use a super-bounded type as a type annotation and as a type argument, and `A<dynamic>` is super-bounded because `A<Null>` is regular-bounded). A super-bounded type is not allowed to be the type of a fresh instance, so the error in line 7 is fine.

**Comment by stereotype441** [comment]
https://dart-review.googlesource.com/c/sdk/+/76280 should fix this.

**Comment by eernstg** [comment]
Seems that there is a need to do more: The example in #34623 yields the following errors with `dartanalyzer` version 2.1.0-dev.5.0:
```
Analyzing n000.dart...
  error • 'dynamic' doesn't extend 'A' at n000.dart:11:7 • type_argument_not_matching_bounds
  error • A value of type '<Y extends A<A<dynamic>>>() → void' can't be assigned to a variable of type '<Y extends A<dynamic>>() → void' at n000.dart:11:27 • invalid_assignment
  error • 'A' doesn't extend 'A<A>' at n000.dart:12:7 • type_argument_not_matching_bounds
  error • 'dynamic' doesn't extend 'A' at n000.dart:12:9 • type_argument_not_matching_bounds
  error • 'A<A>' doesn't extend 'A<A<A>>' at n000.dart:13:7 • type_argument_not_matching_bounds
  error • 'A' doesn't extend 'A<A>' at n000.dart:13:9 • type_argument_not_matching_bounds
  error • 'dynamic' doesn't extend 'A' at n000.dart:13:11 • type_argument_not_matching_bounds
  error • A value of type '<Y extends A<A<dynamic>>>() → void' can't be assigned to a variable of type '<Y extends A<A<A<dynamic>>>>() → void' at n000.dart:13:33 • invalid_assignment
  error • 'A<A<A>>' doesn't extend 'A<A<A<A>>>' at n000.dart:14:7 • type_argument_not_matching_bounds
  error • 'A<A>' doesn't extend 'A<A<A>>' at n000.dart:14:9 • type_argument_not_matching_bounds
  error • 'A' doesn't extend 'A<A>' at n000.dart:14:11 • type_argument_not_matching_bounds
  error • 'dynamic' doesn't extend 'A' at n000.dart:14:13 • type_argument_not_matching_bounds
  error • A value of type '<Y extends A<A<dynamic>>>() → void' can't be assigned to a variable of type '<Y extends A<A<A<A<dynamic>>>>>() → void' at n000.dart:14:36 • invalid_assignment
13 errors found.
```
The error "'dynamic' doesn't extend 'A'" at line 11 shows that the analyzer rejects the type `F<A<dynamic>>`, but this is an application of a parameterized type alias `F` (whose type parameter has no bound) to a type `A<dynamic>`, and `A<dynamic>` is a correct super-bounded type (involving an F-bound, which makes it a bit more involved than the example used in this issue). I think all the other errors containing "doesn't extend" arise for the same reason (so they are also examples of the bug reported in this issue).

The errors containing "can't be assigned" are correct.

**Comment by eernstg** [comment]
Just checked with version 2.1.0-dev.6.0 as well: Same result.

@stereotype441, I don't know whether you'd prefer to create a new issue for this ("super-bounded types involving an F-bound") or to reopen this one, or maybe it's a different person who will look into it, but it would be great if you would take the next step!

**Comment by stereotype441** [comment]
The fix was in https://dart-review.googlesource.com/c/sdk/+/76280, which landed as b1a54154f83c3267628be6d4892e24cba8662ba7.  2.1.0-dev.6.0 was based on c23a9815ec071ece905740801fd1499bcab27729, which was before the fix landed.

I double checked the test case from #34623 with both b1a54154f83c3267628be6d4892e24cba8662ba7 and the latest bleeding edge CL (e90f1f21b8a80d7723d9ed046b8cf38f74072733), and in both cases the analyzer's error messages are:

```
Analyzing /usr/local/google/home/paulberry/tmp/test.dart...
  error • A value of type '<Y extends A<A<dynamic>>>() → void' can't be assigned to a variable of type '<Y extends A<dynamic>>() → void' at /usr/local/google/home/paulberry/tmp/test.dart:9:27 • invalid_assignment
  error • A value of type '<Y extends A<A<dynamic>>>() → void' can't be assigned to a variable of type '<Y extends A<A<A<dynamic>>>>() → void' at /usr/local/google/home/paulberry/tmp/test.dart:11:33 • invalid_assignment
  error • A value of type '<Y extends A<A<dynamic>>>() → void' can't be assigned to a variable of type '<Y extends A<A<A<A<dynamic>>>>>() → void' at /usr/local/google/home/paulberry/tmp/test.dart:12:36 • invalid_assignment
  hint • The value of the local variable 'target1' isn't used at /usr/local/google/home/paulberry/tmp/test.dart:9:17 • unused_local_variable
  hint • The value of the local variable 'target2' isn't used at /usr/local/google/home/paulberry/tmp/test.dart:10:20 • unused_local_variable
  hint • The value of the local variable 'target3' isn't used at /usr/local/google/home/paulberry/tmp/test.dart:11:23 • unused_local_variable
  hint • The value of the local variable 'target4' isn't used at /usr/local/google/home/paulberry/tmp/test.dart:12:26 • unused_local_variable
3 errors and 4 hints found.
```

The only errors that remain are the "can't be assigned" errors, which as you noted above are correct.  So I believe the fix is correct, we just need to wait for it to get released to a dev version.
